File: Data Structure & Algorithm/Review/Stage 10/Review 1/Part1/kruskal/kruskal.py
class DisjointSet:
    def __init__(self, size) -> None:
        self.root = [i for i in range(size)]

    def find(self, x):
        if self.root[x] == x:
            return x
        return self.find(self.root[x])

    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x != root_y:
            self.root[root_y] = root_x

    def isConnected(self, x, y):
        return self.find(x) == self.find(y)


def kruskal(edges: list[list[int]], points: int) -> tuple[int, list[list[int]]]:
    mst: list[list[int]] = []
    cost = 0

    # init
    count = points - 1
    uf = DisjointSet(points)
    edges.sort(key=lambda x: x[2])

    while count > 0:
        v, w, weight = edges.pop(0)

        if not uf.isConnected(v, w):
            mst.append([v, w])
            cost += weight
            count -= 1
            uf.union(v, w)

    return cost, mst


# test
res = kruskal(
    [
        [0, 1, 4],
        [0, 2, 13],
        [0, 3, 7],
        [0, 4, 7],
        [1, 2, 9],
        [1, 3, 3],
        [1, 4, 7],
        [2, 3, 10],
        [2, 4, 14],
        [3, 4, 4],
    ],
    5,
)
print(res)

File: Data Structure & Algorithm/Review/Stage 10/Review 1/Part1/kruskal/kruskal.go
package main

import (
	"fmt"
	"sort"
)

type DisjointSet struct {
	root []int
}

func New(size int) *DisjointSet {
	root := make([]int, size)
	for i := 0; i < size; i++ {
		root[i] = i
	}

	return &DisjointSet{
		root: root,
	}
}

func (uf *DisjointSet) Find(x int) int {
	if x == uf.root[x] {
		return x
	}
	return uf.Find(uf.root[x])
}

func (uf *DisjointSet) Union(x int, y int) {
	rootX := uf.Find(x)
	rootY := uf.Find(y)

	if rootX != rootY {
		uf.root[rootY] = rootX
	}
}

func (uf *DisjointSet) IsConnected(x int, y int) bool {
	return uf.Find(x) == uf.Find(y)
}

func kruskal(graph [][]int, vertices int) (int, [][]int) {
	mst := make([][]int, 0, vertices-1)
	cost := 0

	uf := New(vertices)
	count := vertices - 1
	sort.Slice(graph, func(i, j int) bool {
		return graph[i][2] < graph[j][2]
	})

	for count > 0 {
		current := graph[0]
		graph = graph[1:]

		u, v, weight := current[0], current[1], current[2]

		if !uf.IsConnected(u, v) {
			cost += weight
			mst = append(mst, []int{u, v})
			uf.Union(u, v)
			count--
		}
	}

	return cost, mst
}

// test
func main() {
	cost, mst := kruskal(
		[][]int{
			{0, 1, 4},
			{0, 2, 13},
			{0, 3, 7},
			{0, 4, 7},
			{1, 2, 9},
			{1, 3, 3},
			{1, 4, 7},
			{2, 3, 10},
			{2, 4, 14},
			{3, 4, 4},
		},
		5,
	)
	fmt.Println(cost, mst)
}
